### PR TITLE
Bug 1903999: Httplog response code is always zero

### PR DESCRIPTION
**Pull request by p0lyn0mial**
It turned out that httplog was suddenly disabled for `health checks on the paths like /healthz or /readyz`.

I suspect it was disabled because by default we attach a stack trace for requests with >= 500 HTTP status code which can be verbose (an example below).

I decided to expose a function that allows for disabling putting a stack trace into to log only for `/readyz` handler. Since an HTTP `500` is considered a normal response to signal a LB to stop sending traffic for that path. 

A response with a stack trace:
```
 httplog.go:94] "HTTP" verb="GET" URI="/readyz" latency="3.53127ms" userAgent="ELB-HealthChecker/2.0" srcIP="10.0.91.198:58092" resp=500 statusStack="\ngoroutine 286383 [running]:\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc00ec44ee0, 0x1f4)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:248 +0xcf\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc00ec44ee0, 0x1f4)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:227 +0x35\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).WriteHeader(0xc00ea69ec0, 0x1f4)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:226 +0xb2\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*ResponseWriterDelegator).WriteHeader(0xc0375840f0, 0x1f4)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:571 +0x45\nnet/http.Error(0x5669c60, 0xc00937be40, 0xc0375daa00, 0x625, 0x1f4)\n\t/usr/lib/golang/src/net/http/server.go:2066 +0x1f6\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/healthz.handleRootHealth.func1(0x5669c60, 0xc00937be40, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/healthz/healthz.go:258 +0x64b\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentHandlerFunc.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:474 +0x2ba\nnet/http.HandlerFunc.ServeHTTP(0xc00f75fae0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc02dd9db40, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:241 +0x6fa\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc004916070, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0x8c\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x4da0ead, 0xf, 0xc0009e3050, 0xc004916070, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:154 +0x87f\nk8s.io/kubernetes/openshift-kube-apiserver/openshiftkubeapiserver.withOAuthInfo.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go:54 +0x6f\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd530, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackCompleted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:95 +0x165\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd560, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:64 +0x59a\nnet/http.HandlerFunc.ServeHTTP(0xc00a251b80, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackStarted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:71 +0x186\nnet/http.HandlerFunc.ServeHTTP(0xc00a251bc0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackCompleted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:95 +0x165\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd5f0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8d00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1.4()\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:127 +0x3c6\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol.(*configController).Handle.func1()\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:122 +0x15e\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset.(*request).Finish(0xc029e20bb0, 0xc037920b40, 0xc0173a8390)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:319 +0x42\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol.(*configController).Handle(0xc0009246e0, 0x56704a0, 0xc014d41fb0, 0xc029e20b00, 0x56711a0, 0xc037188e80, 0xc0173a82a0, 0xc0173a82b0, 0xc037920a20)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:115 +0x7aa\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:130 +0x5c3\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd680, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackStarted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:71 +0x186\nnet/http.HandlerFunc.ServeHTTP(0xc00a251c00, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackCompleted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:95 +0x165\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd710, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:50 +0x23e6\nnet/http.HandlerFunc.ServeHTTP(0xc00a251c40, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackStarted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:71 +0x186\nnet/http.HandlerFunc.ServeHTTP(0xc00a251c80, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackCompleted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:95 +0x165\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd7a0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAudit.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit.go:54 +0x750\nnet/http.HandlerFunc.ServeHTTP(0xc00a251cc0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackStarted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:71 +0x186\nnet/http.HandlerFunc.ServeHTTP(0xc00a251d00, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackCompleted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:95 +0x165\nnet/http.HandlerFunc.ServeHTTP(0xc0045dd8c0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8c00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8a00)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:70 +0x6d2\nnet/http.HandlerFunc.ServeHTTP(0xc00e13fa90, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8a00)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency.trackStarted.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8900)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:80 +0x38a\nnet/http.HandlerFunc.ServeHTTP(0xc00a251d80, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8900)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithCORS.func1(0x7f0695d9e438, 0xc00937be20, 0xc0371f8900)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/cors.go:75 +0x1d9\nnet/http.HandlerFunc.ServeHTTP(0xc00306aba0, 0x7f0695d9e438, 0xc00937be20, 0xc0371f8900)\n\t/usr/lib/golang/src/net/http/server.go:2054 +0x44\nk8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc0385796e0, 0xc006615960, 0x56712a0, 0xc00937be20, 0xc0371f8900)\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:111 +0xb8\ncreated by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP\n\t/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:97 +0x1cc\n" addedInfo="\nlogging error output: \"[+]ping ok\\n[+]log ok\\n[+]etcd ok\\n[+]api-openshift-apiserver-available ok\\n[+]api-openshift-oauth-apiserver-available ok\\n[+]informer-sync ok\\n[+]poststarthook/openshift.io-StartOAuthInformers ok\\n[+]poststarthook/start-kube-apiserver-admission-initializer ok\\n[+]poststarthook/quota.openshift.io-clusterquotamapping ok\\n[+]poststarthook/openshift.io-startkubeinformers ok\\n[+]poststarthook/openshift.io-openshift-apiserver-reachable ok\\n[+]poststarthook/openshift.io-oauth-apiserver-reachable ok\\n[+]poststarthook/openshift.io-TokenTimeoutUpdater ok\\n[+]poststarthook/generic-apiserver-start-informers ok\\n[+]poststarthook/priority-and-fairness-config-consumer ok\\n[+]poststarthook/priority-and-fairness-filter ok\\n[+]poststarthook/start-apiextensions-informers ok\\n[+]poststarthook/start-apiextensions-controllers ok\\n[+]poststarthook/crd-informer-synced ok\\n[+]poststarthook/bootstrap-controller ok\\n[+]poststarthook/rbac/bootstrap-roles ok\\n[+]poststarthook/scheduling/bootstrap-system-priority-classes ok\\n[+]poststarthook/priority-and-fairness-config-producer ok\\n[+]poststarthook/start-cluster-authentication-info-controller ok\\n[+]poststarthook/aggregator-reload-proxy-client-cert ok\\n[+]poststarthook/start-kube-aggregator-informers ok\\n[+]poststarthook/apiservice-registration-controller ok\\n[+]poststarthook/apiservice-status-available-controller ok\\n[+]poststarthook/apiservice-wait-for-first-sync ok\\n[+]poststarthook/kube-apiserver-autoregistration ok\\n[+]autoregister-completion ok\\n[+]poststarthook/apiservice-openapi-controller ok\\n[-]shutdown failed: reason withheld\\nreadyz check failed\\n\"\n"
```

a response without the stack trace:
```
I1216 15:41:43.108966      17 httplog.go:94] "HTTP" verb="GET" URI="/readyz" latency="4.899291ms" userAgent="ELB-HealthChecker/2.0" srcIP="10.0.221.206:22463" resp=500
```

TODO:
- check `watch` endpoint

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go [hunk 1]
```diff
@@ -140,10 +140,11 @@ func InstallReadyzHandler(mux mux, checks ...HealthChecker) {
 	InstallPathHandler(mux, "/readyz", checks...)
 }
 
-// InstallReadyzHandlerWithHealthyFunc is like InstallReadyzHandler, but in addition call firstTimeReady
-// the first time /readyz succeeds.
-func InstallReadyzHandlerWithHealthyFunc(mux mux, firstTimeReady func(), checks ...HealthChecker) {
-	InstallPathHandlerWithHealthyFunc(mux, "/readyz", firstTimeReady, checks...)
+// InstallReadyzHandlerWithHealthyFunc is like InstallReadyzHandler but allows for small customization
+// - calls firstTimeHealthy the first time /readyz succeeds.
+// - updating stacktracePred for httplog so that it doesn't log stack trace when HTTP 500 response is returned
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:145]: what does "updating" mean?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 2]
```diff
@@ -135,6 +135,16 @@ func Unlogged(req *http.Request, w http.ResponseWriter) http.ResponseWriter {
 	return w
 }
 
+// UpdateStackTracePred allows you to change the default predicate that is
+// used to decide whether to put a stack trace of the calling goroutine into the log
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:139]: where is that stored? I only see a temporary rl below.

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:145]: updating here means changing behaviour, see https://github.com/openshift/kubernetes/pull/494/files#diff-3d248bbad9c317fb506e12758c05718c9b045dbe0c5b941b0e6bb77ac055c0e7R260

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:139]: `rl` is a pointer to what is stored in the context. We are changing the original object.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:139]: haven't found an easy way to update `req` with `req.WithContext` and pass it further down the chain.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:139]: DisableStackTraceForRequest ?

File: staging/src/k8s.io/apiserver/pkg/server/healthz.go [hunk 3]
```diff
@@ -101,9 +101,14 @@ func (s *GenericAPIServer) installReadyz() {
 	s.readyzLock.Lock()
 	defer s.readyzLock.Unlock()
 	s.readyzChecksInstalled = true
-	healthz.InstallReadyzHandlerWithHealthyFunc(s.Handler.NonGoRestfulMux, func() {
-		// note: InstallReadyzHandlerWithHealthyFunc guarantees that this is called only once
+	healthz.InstallReadyzHandlerWithCustomFunc(s.Handler.NonGoRestfulMux, func() {
+		// note: InstallReadyzHandlerWithCustomFunc guarantees that this is called only once
 		close(s.hasBeenReadyCh)
+	}, func(_ int) bool {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/healthz.go:107]: is there a readyz handler ever where we want stacktraces?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz.go:107]: do you want to default it to `false` ?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:139]: ok, ic, we could do that and drop https://github.com/openshift/kubernetes/pull/494/files#r556379143

File: staging/src/k8s.io/apiserver/pkg/server/healthz.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/healthz.go:107]: yes, without the parameter

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz.go:107]: ok, I'll update the PR. Thanks.

File: staging/src/k8s.io/apiserver/pkg/server/healthz.go [hunk 4]
```diff
@@ -101,8 +101,8 @@ func (s *GenericAPIServer) installReadyz() {
 	s.readyzLock.Lock()
 	defer s.readyzLock.Unlock()
 	s.readyzChecksInstalled = true
-	healthz.InstallReadyzHandlerWithHealthyFunc(s.Handler.NonGoRestfulMux, func() {
-		// note: InstallReadyzHandlerWithHealthyFunc guarantees that this is called only once
+	healthz.InstallReadyzHandlerWithCustomFunc(s.Handler.NonGoRestfulMux, func() {
+		// note: InstallReadyzHandlerWithCustomFunc guarantees that this is called only once
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/healthz.go:105]: undo the rename?

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go [hunk 5]
```diff
@@ -255,8 +257,14 @@ func handleRootHealth(name string, firstTimeHealthy func(), checks ...HealthChec
...
 		if len(failedChecks) > 0 {
+			// update the stacktrace predicate so that httplog's output is condensed
+			// this handler is used for health checking on the paths like /healthz or /readyz
+			// for some (/readyz) the StatusInternalServerError is considered as a "normal" response code to signal the other end
+			if disableStacktraceForHttpLog {
+				httplog.DisableStackTraceForRequest(r)
+			}
 			klog.V(2).Infof("%s check failed: %s\n%v", strings.Join(failedChecks, ","), name, failedVerboseLogOutput.String())
-			http.Error(httplog.Unlogged(r, w), fmt.Sprintf("%s%s check failed", individualCheckOutput.String(), name), http.StatusInternalServerError)
+			http.Error(w, fmt.Sprintf("%s%s check failed", individualCheckOutput.String(), name), http.StatusInternalServerError)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:267]: why this change?

File: staging/src/k8s.io/apiserver/pkg/server/healthz.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz.go:105]: I could but now it is more like `WithHealthyFuncAndDisabledStacktrace`  but OK.

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go (hunk 5 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:267]: to capture the status code and put it in the log file.

`httplog.Unlogged` returns the original ResponseWriter, or w if it is not our inserted logger.

File: staging/src/k8s.io/apiserver/pkg/server/healthz.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz.go]: done.

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go [hunk 6]
```diff
@@ -160,12 +161,13 @@ func InstallLivezHandler(mux mux, checks ...HealthChecker) {
...
+	InstallPathHandlerWithCustomisation(mux, path, nil, false, checks...)
 }
 
-// InstallPathHandlerWithHealthyFunc is like InstallPathHandler, but calls firstTimeHealthy exactly once
-// when the handler succeeds for the first time.
-func InstallPathHandlerWithHealthyFunc(mux mux, path string, firstTimeHealthy func(), checks ...HealthChecker) {
+// InstallPathHandlerWithHealthyFunc is like InstallPathHandler but:
+// - calls firstTimeHealthy exactly once when the handler succeeds for the first time
+// - allows for disabling putting a stacktrace for httplog for the current request so that the output is condensed
+func InstallPathHandlerWithCustomisation(mux mux, path string, firstTimeHealthy func(), disableStacktraceForHttpLog bool, checks ...HealthChecker) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:170]: let's not change the name at all. It keeps the carry change small, and I think there is not much value in `WithCustomization`.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go]: done.
